pkg/operator/controller.go

```diff
@@ -777,25 +775,36 @@ func (c *Controller) getProviderName(nodeName string) (string, error) {
 	return util.GetProviderName(node.Spec.ProviderID), nil
 }
 
-func (c *Controller) syncMachineConfig(pools []*mcfgv1.MachineConfigPool, labels map[string]string, profile *tunedv1.Profile) error {
+func (c *Controller) syncMachineConfig(labels map[string]string, profile *tunedv1.Profile) error {
 	var (
+		bootcmdline     string
 		kernelArguments []string
 	)
 
-	name := GetMachineConfigNameForPools(pools)
-	nodes, err := c.pc.getNodesForPool(pools[0])
+	pools, err := c.pc.getPoolsForMachineConfigLabels(labels)
 	if err != nil {
 		return err
 	}
 
-	bootcmdline, bootcmdlineSet := c.pc.state.bootcmdline[profile.Name]
-	if !bootcmdlineSet {
-		klog.V(2).Infof("syncMachineConfig(): bootcmdline for %s not cached, sync canceled", profile.Name)
-		return nil
-	}
+	name := GetMachineConfigNameForPools(pools)
+	klog.V(2).Infof("syncMachineConfig(): %v", name)
 
-	if ok := c.allNodesAgreeOnBootcmdline(nodes); !ok {
-		return fmt.Errorf("not all %d Nodes in MCP %v agree on bootcmdline: %s", len(nodes), pools[0].ObjectMeta.Name, bootcmdline)
+	for _, pool := range pools {
+		var bootcmdlineSet bool
+		nodes, err := c.pc.getNodesForPool(pool)
+		if err != nil {
+			return err
+		}
+
+		bootcmdline, bootcmdlineSet = c.pc.state.bootcmdline[profile.Name]
+		if !bootcmdlineSet {
+			klog.V(2).Infof("syncMachineConfig(): bootcmdline for %s not cached, sync canceled", profile.Name)
+			return nil
+		}
+
+		if ok := c.allNodesAgreeOnBootcmdline(nodes); !ok {
```

Comment:
If you want to support multiple pools matching the recommendation section of a single Tuned, then I think this loop has to extend further below this. Because you fail the whole sync on two MCPs not having the same kargs. Which they are not required to have, that condition only applies to each MCP separately.

Reply:
> If you want to support multiple pools matching the recommendation section of a single Tuned

@MarSik , I believe there is a fundamental misunderstanding what this change does.  We are not adding support, we are fixing a potential cyclic reboot issue that was not fixed by previous code.

So my aim is not to support multiple MCPs, but we need to check all the matching MCPs affected by this.  Would logging an error or warning when we see multiple matching MCPs be sufficient?

Reply:
But it essentially does add the support for multiple MCPs.

You can block the sync when multiple MCPs are detected instead to make it clear that is is not supported. Not need to validate kargs in such case.

Reply:
> But it essentially does add the support for multiple MCPs.

Let's agree to disagree.  It fixes a bug.

> You can block the sync when multiple MCPs are detected instead to make it clear that is is not supported. Not need to validate kargs in such case.

No, we still need to check the kargs.  You can have machines with different topology within a single MCP.

Reply:
@MarSik , please see the "[Enforce per-pool machineConfigLabels selectors](https://github.com/openshift/cluster-node-tuning-operator/pull/875/commits/2963f6671d6c6f7672dc20b7e114e6abab9af1e9)" commit.  Hopefully it resolves the issues you see with this PR.